Add filter for REST API menu read access check

spacedmonkey opened the pull request:
Introduce 'rest_menu_read_access' filter to allow customization of read access permissions in REST API endpoints for menus, menu items, and menu locations. This facilitates more flexible access control based on custom criteria.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/54304
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Add filter for REST API menu read access check

Introduce 'rest_menu_read_access' filter to allow customization of read access permissions in REST API endpoints for menus, menu items, and menu locations. This facilitates more flexible access control based on custom criteria.

spacedmonkey committed:
Refactor permission checks in menu locations controller

Consolidate permission check logic into a new protected method `check_has_read_only_access` in the `WP_REST_Menu_Locations_Controller` class. This improves code readability and reduces redundancy by centralizing the permission validation process.

anton-vlasenko reviewed (approved):
This looks good to me, but I see the PHP 8.4 tests are failing (seems to be unrelated to this PR).

spacedmonkey committed:
Add tests for menu-related REST API read access filters

Implemented unit tests for the REST API menu endpoints to ensure proper read access behavior. This includes tests for both getting single items and lists of items for menus, menu locations, and menu items. Applied relevant filters and verified response data for accuracy.

kadamwhite commented on tests/phpunit/tests/rest-api/wpRestMenuLocationsController.php:
```diff
@@ -120,6 +120,41 @@ public function test_get_item() {
+		$this->assertCount( 2, $data );
+		$names        = wp_list_pluck( $data, 'name' );
+		$descriptions = wp_list_pluck( $data, 'description' );
+		$this->assertSame( $menus, $names );
```
Nitpick, could we add some linebreaks after the assertions in test_get_items_filter? This becomes hard to read with all the alignment changes, to me at least. Test readability obviously isn't top priority but it'd be a small improvement in my opinion.

spacedmonkey committed:
Merge branch 'trunk' into fix/menu-public

spacedmonkey committed:
Enhance assertions in REST API menu locations tests

Improve PHPUnit test messages for better clarity when assertions fail. Added custom error messages to `assertCount` and `assertSame` methods in menu locations tests. These changes aid in debugging by providing more descriptive feedback.
